### ISSUE TITLE
Feedback for documentation inconsistencies and unnoticed bugs

### ISSUE THREAD

**Issue by stefanos82** [edubart/nelua-lang#68]
In [Global Symbol](https://github.com/edubart/nelua-lang/blob/master/docs/pages/overview.md#global-symbol) of documentation, we have the following example:

```lua
global function global_f()
end
```
This throws an error when called:

```
useglobals.nelua:3:6: error: cannot handle type "void" in print
print(global_f())   -- outputs: f
     ^~~~~~~~~~~~


shell returned 1
```

If I add `return 'f'` inside `global_f()`, I get the desired output.

What is the behavior we initially expected in the first place?

**Comment by edubart** [comment]
Because `global_f` returns nothing, so you can print nothing, you must return at least one value to be able to print. Maybe you expected return `nil` instead  but the Nelua type system at the moment does not convert empty returns to `nil`, but this may change in the future so some lua style codes are more compatible

**Comment by stefanos82** [comment]
I guess the documentation needs a bit of clarification, because in the screenshot below this is what we currently say:

![image](https://user-images.githubusercontent.com/1777466/120845646-c7e01c00-c579-11eb-8017-2282d82608d8.png)

I just follow the documentation to understand the language in detail.

**Comment by edubart** [comment]
Oh, its a mistake in the documentation.

**Comment by stefanos82** [comment]
I see. Well, just add a `return 'f'` inside the function and it will print the desired output.

**Comment by stefanos82** [comment]
By the way @edubart can we transfer this to Discussion? I will be testing the whole docs and in case I find other buggy code samples, I will be reporting it there instead of opening multiple tickets.

Is that OK with you?

**Comment by edubart** [comment]
Yes, I am ok with this, continue posting here any mistake you find in the documentation.

**Comment by stefanos82** [comment]
Thank you @edubart

**Comment by stefanos82** [comment]
![image](https://user-images.githubusercontent.com/1777466/120890917-67e78500-c60e-11eb-85cb-18da140d7d8c.png)

In reality it prints `false	test	1	3.14`

**Comment by stefanos82** [comment]
In `condots.nelua` we have the following code:
```lua
local function poll_events()
  local event: SDL_Event
  while SDL_PollEvent(event) ~= 0 do
    switch event.type
    case SDL_QUIT then
      return false
    end
  end
  return true
end
```
but in docs we have

![image](https://user-images.githubusercontent.com/1777466/120903094-d8ad9200-c64c-11eb-8f7c-b87c9d88fae4.png)

Is `do` optional in this case or is it a bug? The code compiles and works as expected :man_shrugging:

**Comment by edubart** [comment]
> Is `do` optional in this case or is it a bug? The code compiles and works as expected

The `do` keyword following the switch expression is currently optional at the moment. At first it wasn't possible to use `do`, but a user suggested to add the `do` keyword, so it was added just as an optional syntax sugar.

**Comment by edubart** [comment]
I've fixed the following things you mentioned in the documentation:
* Missing the `return 'f'` in global symbol section
* 3.14 output instead of 3.1400000 in variables section
* Omitted the `do` keyword in switch statement (using the `do` is optional and will remain as optional hidden feature, I won't document this possibility as this may be removed in the future to not have extra syntax).

Thanks for reporting these inconsistencies, once you feel there is nothing more wrong in the documentation you can close this issue.

**Comment by stefanos82** [comment]
Sure thing @edubart, I currently work on implementing syntax highlighting for Vim editor which I use and it's a nightmare :smile: that's why I report all these inconsistencies; I'm trying to make it work as expected.

As soon as I feel it's ready I will create a repo around it for people to use it and provide feedback accordingly.

**Comment by edubart** [comment]
> Sure thing @edubart, I currently work on implementing syntax highlighting for Vim editor which I use and it's a nightmare  that's why I report all these inconsistencies; I'm trying to make it work as expected.
> 
> As soon as I feel it's ready I will create a repo around it for people to use it and provide feedback accordingly.

If you are working on a syntax highlighter plugin I highly recommend using the compiler PEG grammar as reference that just shipped the master branch a few hours ago. This file:

https://github.com/edubart/nelua-lang/blob/master/nelua/syntaxdefs.lua#L14

You can read the full syntax definition in a compact form there as reference (assuming you understand LPEG's re syntax).

**Comment by stefanos82** [comment]
Very nice! I will double-check everything right now; thank you for the feedback.

Vimscript's syntax is pure...magic? Yeah, I think you understand what I mean :expressionless:

**Comment by stefanos82** [comment]
I have found two types that are undocumented: `union` and `variant`.

**Comment by edubart** [comment]
> I have found two types that are undocumented: `union` and `variant`.

Variant is a planned type but not implemented yet so it can't be in the documentation yet, because the documentation only have working features. Union type is new, it is implemented mainly for importing C unions, but need more testing and polishiments before adding it officially to the documentation.

**Comment by stefanos82** [comment]
I see. By the way, the PEG syntax is very clean; it will be handy for me now that I'm learning compilers and programming language design. It's an incredible topic that always interested me.

**Comment by stefanos82** [comment]
![image](https://user-images.githubusercontent.com/1777466/120908963-6fdc0f00-c678-11eb-82bb-0b37162735d2.png)

Did you mean `a_scalar` here or `an_` means something that gets interpreted during preprocessing?

**Comment by stefanos82** [comment]
OK, this is very interesting @edubart...

I copied the following code from `examples/overview.lua` and expected to get the commented output:

```lua

do
  local a = 1234
  local b = 0xff
  local c = 3.14159
  local d: integer
  print(a,b,c,d) -- outputs: 1234 255 3.141590 0

  do
    local a = 1234_u32
    local b = 1_f32
    print(a,b) --outputs: 1234 1.000000
  end
end
```
This was the output instead:

```
1234	255	3.14159	0
1234	1
```
The culprit in this problematic output is in the generated C code, that is the use of `%.14g` and `%.7g` respectively.

**Comment by edubart** [comment]
> The culprit in this problematic output is in the generated C code, that is the use of `%.14g` and `%.7g` respectively.

You have found a bug! Thanks for reporting. This has been fixed in the latest commit. The `print` function should behave like Lua, and always print floats removing trailing zeroes, but always add at least `.0` in case the float is not fractional. This inconsistency was fixed in the code and documentation.

> Did you mean a_scalar here or an_ means something that gets interpreted during preprocessing?

I know `an_scalar` is grammatically incorrect there, but this kind of concept is used other places like `an_integer`, `an_array`. To make things more regular I opted to prefix all concepts like that one with `an_`, instead of making each one grammatically correct, though I may rethink this style later.

**Comment by stefanos82** [comment]
> You have found a bug! Thanks for reporting. This has been fixed in the latest commit. The `print` function should behave like Lua, and always print floats removing trailing zeroes, but always add at least `.0` in case the float is not fractional. This inconsistency was fixed in the code and documentation.

Nice! I hope I find more things that need to get fixed so you can get closer to release a stable version of Nelua.

> I know `an_scalar` is grammatically incorrect there, but this kind of concept is used other places like `an_integer`, `an_array`. To make things more regular I opted to prefix all concepts like that one with `an_`, instead of making each one grammatically correct, though I may rethink this style later.

Well, the aforementioned names start with a vowel, that's why it makes sense to use `an_`.

**Comment by stefanos82** [comment]
By the way, what is the way to print an address of a variable in Lua or Nelua?

**Comment by stefanos82** [comment]
I compiled the latest commit and I still don't get the same output as this particular code:

![image](https://user-images.githubusercontent.com/1777466/120923558-e4926600-c6d7-11eb-89c8-0141fc413d82.png)

`1234	255	3.14159	0`

The second `do end` works as expected.

**Comment by edubart** [comment]
> I compiled the latest commit and I still don't get the same output as this particular code

Oh just forgot to remove that extra `0` in the comment, that typo be fixed in the next interesting commit.

> By the way, what is the way to print an address of a variable in Lua or Nelua?

If a variable is a pointer you can just print, if it's not you can get its pointer with the reference operator `&`, example:
```lua
local i: integer -- outputs something like 0x55a9b71b7058
print(&i)
local p: *integer
print(p) -- outputs: (null)
p = &i
print(p) -- outputs something like 0x55a9b71b7058
```

**Comment by stefanos82** [comment]
Loved it! I'm taking notes :laughing:

**Comment by edubart** [comment]
> I have found two types that are undocumented: `union` and `variant`.

I've added union to the docs, in short the following new sections:
* Union
* Multiple variables assignment
* For In
* Variable number of arguments
* Multidimensional Array

**Comment by stefanos82** [comment]
I have tried the example of _Multiple variable assignment_ and decided to assign it to a single variable only.

Should not there be a warning for such thing? The second returned value is silenced I presume...

**Comment by edubart** [comment]
> Should not there be a warning for such thing? The second returned value is silenced I presume...

This is not a warning because this is valid in Lua, thus is also valid Nelua so both can have similar semantics.
To be clear when declaring two variables while assigning just one the second variable will be initialized to `nil` like in Lua.
Although this is not much of use at this time, as you cant do much with a variable of type `nil`, but it will be usable in the future.

**Comment by stefanos82** [comment]
Well, what I can do is to use the dump variable `_` to discard the values I don't need and that would work as well, like how Golang does: `local x, _ = get_multiple(); print(x)`

**Comment by stefanos82** [comment]
Feedback please...

```lua
for i = 0,5 do
  -- i is deduced to 'integer'
  print(i) -- outputs 0 1 2 3 4 5
end
```
In the code above we let the user know what the output will look like and we show it in linear order; in reality is like this:
![image](https://user-images.githubusercontent.com/1777466/121020977-8c2b9900-c7a9-11eb-8af4-efd8afc57419.png)

Should we clarify this or we don't care, as long as the output matches the comment section?

**Comment by edubart** [comment]
> Should we clarify this or we don't care, as long as the output matches the comment section?

Well it's just a hint of the output, I don't think there is need to clarify the output with exact word match in the comments at cost of more dense text, I think the reader will get the point.

I am taking the chance and slowly adding missing information to the documentation, added now these sections:

* Do expression
* String escape sequence
* Anonymous functions
* Nested functions

**Comment by stefanos82** [comment]
Yeah, I have noticed it while pulling new commit changes.

**Comment by stefanos82** [comment]
The example from _For In_ shows the following example: `local a: [4]string = {"a","b","c","d"}`

I have looked in the documentation and could not find anything about determining array size upon initialization, like in C for example: `local a: []string = {"a","b","c","d"}`

I have tested the aforementioned code and works.

Is it undocumented code or a feature?

**Comment by edubart** [comment]
> Is it undocumented code or a feature?

It was undocumented, but now it is documented.

**Comment by stefanos82** [comment]
Awesome!

Which types `auto` supports by default?

I'm asking because the following code kind of confuses me in a way:

```lua
local a: []integer = {1,2,3,4}
print('array size: ',#a) -- outputs: array size:     4

local b: auto = {2.0,3.0}
print(b)
-- throws an error:

inferredarray.nelua:4:7: error: type 'table' is not supported yet in the C backend
local b: auto = {2.0,3.0}
      ^~~~~~~

```

So my actual question should be: why in array type behaves like an array code block whereas with `auto` sees it as an attempt to a not-yet-supported table?

**Comment by edubart** [comment]
> So my actual question should be: why in array type behaves like an array code block whereas with `auto` sees it as an attempt to a not-yet-supported table?

An expression with curly braces `{ ... }` in the compiler is defined as an Initializer List. Initializer lists don't have a fixed type on its own, the compiler needs other type information to decide what type the initializer list should resolve. An initializer list can resolve to an array, a record, an union or a table (not implemented yet, but will be just like Lua tables in the future), it all depends on the variable type on assignment.

For example:
```lua
local Array2 = @[4]integer
local Vec2 = @record{x: integer, y: integer}

local a: Array2 = {1,2} -- the compiler knows that the initializer list is going to be an Array2
local p: Vec2 = {1,2} -- the compiler knows that the initializer list is going to be an Vec2

-- if you uncomment the following, what the type should be inferred? an Array2 or a Vec2? the compiler can't know
--local a: auto = {1, 2}
```

Workaround is to enforce the type on the initializer list expression:
```lua
local a: auto = (@[4]integer){1,2,3,4}
```

So in short, you need a type notation to hint the compiler what the Initializer List will become, otherwise the compiler will assume that is going to be a table.

**Comment by stefanos82** [comment]
> An expression with curly braces `{ ... }` in the compiler is defined as an Initializer List. Initializer lists don't have a fixed type on its own, the compiler needs other type information to decide what type the initializer list should resolve. An initializer list can resolve to an array, a record, an union or a table (not implemented yet, but will be just like Lua tables in the future), it all depends on the variable type on assignment.
>

I see. Since you have mentioned those types, what about `enum`? You left it out.

> For example:
> 
> ```lua
> local Array2 = @[4]integer
> local Vec2 = @record{x: integer, y: integer}
>

Here `Array2` and `Vec2` are considered aliased types, like C's `typedef`, or actual variables?
 
> local a: Array2 = {1,2} -- the compiler knows that the initializer list is going to be an Array2
> local p: Vec2 = {1,2} -- the compiler knows that the initializer list is going to be an Vec2
> 
> -- if you uncomment the following, what the type should be inferred? an Array2 or a Vec2? the compiler can't know
> --local a: auto = {1, 2}
> ```

Neither; it should use the default behavior whichever is when it sees "naked" curly braces, without actual type to explicitly state what it needs to use; thus, Nelua needs one for resolving such limitation. `@record{}` should be the default in my opinion until `table` gets implemented.

Here's the tricky part; the `local a: auto = {1,2}` if I understood Nelua correctly, could match `enum`, `array`, `record`, and logically `union` too.

> 
> Workaround is to enforce the type on the initializer list expression:
> 
> ```lua
> local a: auto = (@[4]integer){1,2,3,4}
> ```

That's an oxymoron, don't you think?

You use `auto` to let the compiler figure out the type for you, yet you type cast `a` to an array of 4 elements of type integer.

> 
> So in short, you need a type notation to hint the compiler what the Initializer List will become, otherwise the compiler will assume that is going to be a table.

Yes, I suggested something like that above.

I will try to find possible solution around this idea just for the sake of keeping my head spinning.

**Comment by edubart** [comment]
> Here Array2 and Vec2 are considered aliased types, like C's typedef, or actual variables?

Yes they are aliases, like C's typedefs, that's in the documentation.

> `@record{}` should be the default in my opinion until table gets implemented.

This is not really possible because initializer lists cannot declare records on its own, plus the language won't keep change behavior due to missing planned features, braces expressions without type notation will remain reserved for the planned tables, which will be a big feature.

> could match enum, array, record, and logically union too.

Not really, braces inside type expressions behave differently, not all braces in the syntax are initializer lists.

> That's an oxymoron, don't you think?

Yes, auto is not there really to be used like this, the workaround was just an example to show how it could work for the problem you mentioned, maybe I should remove auto examples from the documentation, the only reason auto exists is to use in polymorphic function arguments, moreover you won't see much `local x: auto` style in the code.

The documentation already states the following:

> Auto variables were not intended to be used in variable declarations
like in the example above, because in most cases you can omit the type
and the compiler will automatically deduce it.
This can be used, however, if you want the compiler to deduce early.
The `auto` type was mainly created to be used with [polymorphic functions](#polymorphic-functions).

You should keep coding as if `auto` didn't exist, once you start playing with polymorphic functions without concepts you will find where `auto` could be used.

**Comment by stefanos82** [comment]
> Yes they are aliases, like C's typedefs, that's in the documentation.
>

Would it be too much if I kindly ask for introducing a new keyword, `alias`? Because to be honest with you

```lua
local Array2 = @[4]integer
local Vec2 = @record{x: integer, y: integer}
```
looks like declaring variables, whereas

```lua
alias Array2 = @[4]integer
alias Vec2 = @record{x: integer, y: integer}
```
makes it quite obvious I'm defining new type.
 
> > `@record{}` should be the default in my opinion until table gets implemented.
> 
> This is not really possible because initializer lists cannot declare records on its own, plus the language won't keep change behavior due to missing planned features, braces expressions without type notation will remain reserved for the planned tables, which will be a big feature.
> 

Then the current behavior is more than enough; it works more than well in my humble opinion, no need to add an unnecessary layer of complexity.

> Yes, auto is not there really to be used like this, the workaround was just an example to show how it could work for the problem you mentioned, maybe I should remove auto examples from the documentation, the only reason auto exists is to use in polymorphic function arguments, moreover you won't see much `local x: auto` style in the code.
> 

Maybe you should move it to _Polymorphic functions_ section exclusively? That would be an option.

> You should keep coding as if `auto` didn't exist, once you start playing with polymorphic functions without concepts you will find where `auto` could be used.

Yeah, that's what I plan to do.
